### PR TITLE
[flang] fix build on FreeBSD after 4762c6557d15

### LINKED ISSUE #86038: flang failing on build on FreeBSD
I've been attempting to a periodic update of our main-tracking LLVM package in FreeBSD and hitting this compilation error in flang:
```
/usr/bin/c++ -DFLANG_INCLUDE_TESTS=1 -DFLANG_LITTLE_ENDIAN=1 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MAC[35/1859]TDC_LIMIT_MACROS -I/usr/local/poudriere/ports/brooks/devel/llvm-devel/w[34/1859]lt/.build/tools/flang/runtime -I/usr/local/poudriere/ports/brooks/devel[33/1859]el/work-default/llvm-project-e74bcecd36a59a9271c7d4133e73448a3def74c1/f[32/1859]ime -I/usr/local/poudriere/ports/brooks/devel/llvm-devel/work-default/l[31/1859]ct-e74bcecd36a59a9271c7d4133e73448a3def74c1/flang/include -I/usr/local/[30/1859]/ports/brooks/devel/llvm-devel/work-default/.build/tools/flang/include [29/1859]cal/poudriere/ports/brooks/devel/llvm-devel/work-default/.build/include[28/1859]ocal/poudriere/ports/brooks/devel/llvm-devel/work-default/llvm-project-[27/1859]6a59a9271c7d4133e73448a3def74c1/llvm/include -isystem /usr/local/poudri[26/1859]/brooks/devel/llvm-devel/work-default/llvm-project-e74bcecd36a59a9271c7[25/1859]48a3def74c1/llvm/../mlir/include -isystem /usr/local/poudriere/ports/br[24/1859]l/llvm-devel/work-default/.build/tools/mlir/include -isystem /usr/local[23/1859]e/ports/brooks/devel/llvm-devel/work-default/.build/tools/clang/include[22/1859] /usr/local/poudriere/ports/brooks/devel/llvm-devel/work-default/llvm-p[21/1859]4bcecd36a59a9271c7d4133e73448a3def74c1/llvm/../clang/include -O2 -pipe [20/1859]rotector-strong -isystem /usr/local/include -fno-strict-aliasing  -isys[19/1859]local/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hi[18/1859]ror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unu[17/1859]eter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic[16/1859]r-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /[1190/1859]nclude -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wno-deprecated-copy -Wno-string-conversion -Wno-ctad-maybe-unsupported -Wno-unused-command-line-argument -Wstring-conversion           -Wcovered-switch-default -Wno-nested-anon-types -fno-lto -O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /usr/local/include  -DNDEBUG -std=c++17   -U_GLIBCXX_ASSERTIONS  -fno-exceptions -funwind-tables -UNDEBUG -MD -MT tools/flang/lib/Decimal/CMakeFiles/obj.FortranDecimal.dir/decimal-to-binary.cpp.o -MF tools/flang/lib/Decimal/CMakeFiles/obj.FortranDecimal.dir/decimal-to-binary.cpp.o.d -o tools/flang/lib/Decimal/CMakeFiles/obj.FortranDecimal.dir/decimal-to-binary.cpp.o -c /usr/local/poudriere/ports/brooks/devel/llvm-devel/work-default/llvm-project-e74bcecd36a59a9271c7d4133e73448a3def74c1/flang/lib/Decimal/decimal-to-binary.cpp
/usr/local/poudriere/ports/brooks/devel/llvm-devel/work-default/llvm-project-e74bcecd36a59a9271c7d4133e73448a3def74c1/flang/lib/Decimal/decimal-to-binary.cpp:38
1:24: error: called object type 'float' is not a function or function pointer   
      return {Real{HUGE()}};                                                    
                   ~~~~^                                                        
1 error generated.                                                              
```

Host info:
```
$ /usr/bin/c++ --version
FreeBSD clang version 16.0.6 (https://github.com/llvm/llvm-project.git llvmorg-16.0.6-0-g7cbf1a259152)
Target: x86_64-unknown-freebsd14.0
Thread model: posix
InstalledDir: /usr/bin
$ uname -a
FreeBSD hostname 14.0-RELEASE-p3 FreeBSD 14.0-RELEASE-p3 #0: Mon Dec 11 04:56:01 UTC 2023     root@amd64-builder.daemonology.net:/usr/obj/usr/src/amd64.amd64/sys/GENERIC amd64
```

LLVM source is at e74bcecd36a59a9271c7d4133e73448a3def74c1, but this has been going on for a while.

I attempted to bisect and landed on f7a15e0021697e2346d3aa335dedf2bb3cf468f9, but now it doesn't reproduce after a `ninja clean` so I think there's something funky going on in the build infrastructure. I can attempt a bisect with completely clean builds.

### PR DESCRIPTION
**Pull request by brooksdavis**
The HUGE definition collides with the HUGE macro from math.h.  Unlike the fix in 3149c934cb26 (#84478) (largely reverted in f95710c76519), add another #undef HUGE since there is no practical way to make FreeBSD's headers not define HUGE and still define XSI interfaces such as isascii or strnlen.

Fixes #86038

### CODE REVIEW

File: flang/lib/Decimal/decimal-to-binary.cpp [hunk 1]
```diff
@@ -16,6 +16,10 @@
 #include <ctype.h>
 #include <utility>
 
+// Some environments, viz. glibc 2.17, allow the macro HUGE
+// to leak out of <math.h>.
```
Comment (luporl) [inline flang/lib/Decimal/decimal-to-binary.cpp:20]: Please mention in this comment that FreeBSD also allows the macro HUGE to leak out of <math.h>.

**Review by luporl — APPROVED** [review]
If there is no practical way to make FreeBSD's headers not define HUGE and still define XSI interfaces such as isascii or strnlen, then there's not much choice but to undef HUGE.

A possibly better option would be to include `<cmath>` and undef HUGE in `flang/lib/Decimal/big-radix-floating-point.h`, to avoid issues with other sources using its HUGE definition in the future. By doing this in every header file that declares a HUGE symbol, it should be possible to avoid hitting this issue again.

File: flang/lib/Decimal/decimal-to-binary.cpp (hunk 1 above)
Comment (brooksdavis) [reply flang/lib/Decimal/decimal-to-binary.cpp]: I've added a second commit to update all the comments.

**Comment by brooksdavis** [comment]
I'll leave a better fix to the flang developers. It appears that HUGE is de facto part of the math.h interface and POSIX should just define it if the definitions are reasonably well aligned.

**Review by brad0 — APPROVED** [review]

**Pull request merged**